src/derivative_utils.jl

```diff
@@ -167,17 +181,27 @@ function LinearAlgebra.mul!(Y::AbstractVecOrMat, W::WOperator, B::AbstractVecOrM
       mul!(Y, W.mass_matrix, B)
     end
     # Compute J * B
-    mul!(W.cache, W.J, B)
+    mul!(W._func_cache, W.J, B)
     # Subtract result
-    axpy!(-W.gamma, W.cache, Y)
+    axpy!(-W.gamma, W._func_cache, Y)
   end
 end
 
+"""
+    lazy_W(f) -> t/f
+
+Predicate for determining what kind of function supports the use of lazy W.
+Also used in cache construction.
+"""
+lazy_W(f) = false # default
```

Comment:
Can't it always support lazy W? It's non-lazy that not always possible.

Reply:
To use lazy W the `jac_prototype` needs to be an diffeq operator.

Of course I can modify the constructor so that regular `jac` & `jac_prototype` are automatically wrapped in a `DiffEqArrayOperator`, so it's mostly a design choice. If you think this is preferrable then I can make the change (I would need to add DiffEqOperators to the requirement list though; this was actually my reason for going against it, but now that I think about this you need to import DiffEqOperators anyway for the operator interface so this argument doesn't make sense).

There's also the problem of dual mass matrix of `DynamicalFunction`, which I do not know how to handle just yet. For `SplitFunction` I need to probably add a `jac_prototype` field, while for other `*DEFunction` I think it's better to handle them in their respective packages.

Reply:
>There's also the problem of dual mass matrix of DynamicalFunction, which I do not know how to handle just yet.

We don't have any implicit methods for partitioned ODEs, so we can just avoid it for now. I think that for anything like a second order ODE or Hamiltonian problem you'll have that the second mass matrix function has to be I (?), so it may be a non-issue. We'll get to that when we get there.

>For SplitFunction I need to probably add a jac_prototype field, while for other *DEFunction I think it's better to handle them in their respective packages.

Yeah, though remember that StochasticDiffEq.jl will have the same structure as here so we should make sure this same setup gets ported there soon after this merges.